feat: store writer framework IDE into project directory instead of ui.json

FabienArcellier opened the pull request:
implement https://github.com/FabienArcellier/writer-framework/issues/67

FabienArcellier commented on src/writer/core.py:
```diff
@@ -2101,6 +2103,106 @@ def build_writer_func_arguments(func: Callable, writer_args: dict) -> List[Any]:
+    if not os.path.exists(wf_directory):
+        os.makedirs(wf_directory)
+
+    with open(os.path.join(wf_directory, "metadata.json"), "w") as f:
```
use `io.open`

FabienArcellier commented on src/writer/app_runner.py:
```diff
@@ -712,6 +714,9 @@ async def update_components(self, session_id: str, payload: ComponentUpdateReque
+
+        wf_project_write_files(self.app_path, metadata={"writer_version": VERSION}, components=payload.components)
+
         file_contents = {
```
remove this part of the code

FabienArcellier commented on src/writer/core.py:
```diff
@@ -2101,6 +2103,107 @@ def build_writer_func_arguments(func: Callable, writer_args: dict) -> List[Any]:
+    assert os.path.isfile(os.path.join(app_path, "ui.json")), f"ui.json file required for migration into {app_path}"
+
+    logger = logging.getLogger('writer')
+    with open(os.path.join(app_path, "ui.json"), "r") as f:
```
use io.open

FabienArcellier commented on src/writer/core.py:
```diff
@@ -2145,6 +2248,29 @@ def _assert_record_match_list_of_records(df: List[Dict[str, Any]], record: Dict[
 
 
+def _component_page(components: dict[str, ComponentDefinition], component_id: str) -> Optional[str]:
+    """
```
rename into `_lookup_page_for_component`

ramedina86 commented:
This looks good, couple of questions:
- Can we avoid growing the `core`? I don't think there are any dependencies, happy to start a new module for this purpose
- For Workflows, I've decided to start a new root, so we'll have two roots, "root" and "workflows_root". What's the impact of this?

ramedina86 commented:
Let's also update docs at https://dev.writer.com/framework/quickstart#create-an-app

FabienArcellier commented on tests/backend/fixtures/file_fixtures.py:
```diff
@@ -0,0 +1,58 @@
+            return json.loads(content)
+        elif format == FileFormat.jsonl:
+            output = []
+            for line in content.splitlines():
```
`json.dumps` will escape the character `\n`. So, the encoding on one line is preserved

![image](https://github.com/user-attachments/assets/1407ab78-fe52-47e7-a950-44f467eb1572)

FabienArcellier commented on tests/backend/fixtures/components/components-root.jsonl:
```diff
@@ -0,0 +1 @@
+{"id": "root", "type": "root", "content": {}, "isCodeManaged": false, "position": 0}
```
This file is used as fixture to create object for automatic testing. The engine load it properly. I would say it should not be an issue.

![image](https://github.com/user-attachments/assets/ca203b43-f875-4229-8afc-3c8f989d560c)

FabienArcellier commented:
> This looks good, couple of questions:
> 
> * Can we avoid growing the `core`? I don't think there are any dependencies, happy to start a new module for this purpose
> * For Workflows, I've decided to start a new root, so we'll have two roots, "root" and "workflows_root". What's the impact of this?

I moved the methods prefixed with wf_project_ into a new module, `wf_project`. I moved the `lookup_page_for_component` method into the `core_ui` module.

I would prefer that the workflows share the same root and are therefore found in `workflows-*.jsonl` files. At the time of implementation, it will be necessary to change the signature of `wf_project.write_files` and `wf_project.read_files`.

```python
def write_files(app_path: str, metadata: MetadataDefinition, components: dict[str, ComponentDefinition], workflows: [str, WorkflowDefintion]) -> None: ...

def read_files(app_path: str) -> Tuple[MetadataDefinition, dict[str, ComponentDefinition], dict[str, WorkflowDefinition]]: ...
```

ramedina86 commented on docs/framework/custom-server.mdx:
```diff
@@ -7,7 +7,7 @@ Framework uses Uvicorn and serves the app in the root path i.e. `/`. If you need
 
 You can tune your server by adding a `server_setup.py` file to the root 
-of your application, next to the `main.py` and `ui.json` files.
+of your application, next to the `main.py` files.
```
```suggestion
of your application, next to the `main.py` file.
```
